sdk/core/System.ClientModel/src/Message/PipelineResponse.cs

```diff
@@ -0,0 +1,89 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.IO;
+
+namespace System.ClientModel.Primitives;
+
+public abstract class PipelineResponse : IDisposable
+{
+    // TODO(matell): The .NET Framework team plans to add BinaryData.Empty in dotnet/runtime#49670, and we can use it then.
+    private static readonly BinaryData s_emptyBinaryData = new(Array.Empty<byte>());
+
+    private bool _isError = false;
+
+    /// <summary>
+    /// Gets the HTTP status code.
+    /// </summary>
+    public abstract int Status { get; }
+
+    /// <summary>
+    /// Gets the HTTP reason phrase.
+    /// </summary>
+    public abstract string ReasonPhrase { get; }
+
+    public MessageHeaders Headers => GetHeadersCore();
+
+    protected abstract MessageHeaders GetHeadersCore();
+
+    /// <summary>
+    /// Gets the contents of HTTP response. Returns <c>null</c> for responses without content.
+    /// </summary>
+    public abstract Stream? ContentStream { get; set; }
+
+    #region Meta-data properties set by the pipeline.
+
+    public BinaryData Content
+    {
+        get
+        {
+            if (ContentStream == null)
+            {
+                return s_emptyBinaryData;
+            }
+
+            if (!TryGetBufferedContent(out MemoryStream bufferedContent))
```

Comment:
I filed a comment in the other PR: I dont think we should be storing buffered bytes in MemoryStream. The buffered policy should use MemoryStream (or even better SequenceBuilder from Michael's part of the dll) to create the buffered payload and then store it. Our mainline scenario is buffered content and we should optimize for it.

Reply:
Let me see how much work is involved here - if it looks substantial, we can file an issue to track, since it's an internal implementation detail that could feasibly be addressed after the first release.  Offhand, I think there are a lot of Azure.Core tests that depend on response.ContentStream being a MemoryStream to indicate that it has been buffered, but I can bring a more thorough assessment of that.

Reply:
@KrzysztofCwalina, in this solution, PipelineResponse.ContentStream is null if it is not streaming content?  i.e. the ResponseBufferingPolicy sets PipelineResponse.Content with a BinaryData instance and removes the network stream and the intermediate MemoryStream buffer is discarded?

Reply:
If this is correct, I would be concerned about introducing a behavioral breaking change to Azure.Core's Response type where I expect a lot of code has been written with the expectation that Response.ContentStream will be populated with a MemoryStream when the response is buffered.  If wanted to introduce this behavior in PipelineResponse but not Azure.Response, we might be able to rework the Azure.Core buffering policy to save the memory stream somehow, but I would need to look at how that would work since it would mean doing the buffering in the ClientModel policy and then passing the buffer back to the Azure.Core policy somehow.  Alternatively, since we've moved the buffering logic into PipelineResponse, we could just keep those two policies separate - I'll look at how much logic gets duplicated in that case if we do that instead of having the Azure.Core policy call into the ClientModel one.

Either way, let's sync on this next week so I can better understand the picture you have in your head and we can look at the tradeoffs around the different approaches together.

Reply:
Actually, we could pass a bool to the policy telling it whether or not to set the content stream?

Reply:
Ok, I put together a quick prototype PR illustrating how I imagine you're suggesting we make the change, without e.g. SequenceBuilder optimizations -- we can add optimizations if we decide to move forward with the change.  It's here:  https://github.com/Azure/azure-sdk-for-net/pull/41047.  Let me know if that's what you had in mind, and we'll see if I've done the right thing to maintain back-compat with Azure.Core tests if the CI's stay green 🙂

Reply:
Even this implementation does break tests we have today in Azure.Core that are validating that an exception is thrown when the response is not buffered, such as https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/tests/HttpPipelineFunctionalTests.cs#L70.

Given this, I'd like to be more cautious regarding how we make this change, and since it's an internal detail, perhaps we postpone it until after we're confident we have a public API surface we're ready to GA.  I'll file an issue to track doing this work going forward.

Reply:
Tracking with https://github.com/Azure/azure-sdk-for-net/issues/41080